metals/src/main/scala/scala/meta/internal/parsing/JavaFoldingRangeExtractor.scala

```diff
@@ -1,223 +1,230 @@
-            lBrace.line - 1,
-            lBrace.ch,
-            endLine - 1,
-            endCharacter,
-            FoldingRangeKind.Region,
-          ) match {
-            case Some(newFoldingRange) =>
-              group(tail, newFoldingRange :: acc)
-            case None =>
-              group(tail, acc)
-          }
+  }
 
-        case (region: Region) :: tail =>
-          createFoldingRange(
-            region.startLine - 1,
-            region.startCh,
-            region.endLine - 1,
-            region.endCh,
-            region.kind,
-          ) match {
-            case Some(newFoldingRange) =>
-              group(tail, newFoldingRange :: acc)
-            case None =>
-              group(tail, acc)
-          }
+  private def getTrees(
+      text: String,
+      path: AbsolutePath,
+  ): Option[(Trees, CompilationUnitTree)] = {
+    // TODO - surely this can be shared.  Maybe in JavaMetalsGlobal. Or in java equivalent of scala's Trees
+    val javaGlobal = new JavaMetalsGlobal(null, null)
+    val javacTask = javaGlobal.compilationTask(text, path.toURI)
+    val elems = javacTask.parse()
+    javacTask.analyze()
+    val trees = Trees.instance(javacTask)
+    val iter = elems.iterator
+    // only one CompilationUnitTree for a single file
+    if (iter.hasNext) Some((trees, iter.next())) else None
+  }
 
-        case _ =>
-          acc.reverse
+  // search for any comments in file that aren't defined within a String
+  private def findComments(
+      text: String,
+      exclusions: List[Range],
+  ): List[Range] = {
+    @tailrec
+    def findComments(comments: List[Range], idx: Int): List[Range] = {
+      val startIdx = text.indexOf("/*", idx)
+      if (startIdx == -1)
+        comments
+      else {
+        val (newComments, newIdx) =
+          if (exclusions.exists(range => range.contains(startIdx.longValue)))
+            (comments, startIdx + 2)
+          else {
+            val endIdx = text.indexOf("*/", startIdx + 2)
+            val range = Range(
+              startIdx.longValue,
+              endIdx.longValue + 2,
+              FoldingRangeKind.Comment,
+            )
+            (range :: comments, endIdx + 2)
+          }
+        findComments(newComments, newIdx)
       }
-
     }
-    group(startAndEnds, Nil)
+
+    findComments(Nil, 0)
   }
 
-  private def createFoldingRange(
-      startLine: Int,
-      startCharacter: Int,
-      endLine: Int,
-      endCharacter: Int,
-      kind: String,
-  ): Option[FoldingRange] = {
-    if (endLine - startLine >= spanThreshold) {
-      val newFoldingRange = new FoldingRange(startLine, endLine)
-      newFoldingRange.setStartCharacter(startCharacter)
-      newFoldingRange.setEndCharacter(endCharacter)
-      newFoldingRange.setKind(kind)
-      Some(newFoldingRange)
-    } else {
-      None
+  def extract(
+      text: String,
+      path: AbsolutePath,
+      // TODO - what is this supposed to do???  Is it relevant in Java?
```

Comment:
As mentioned above `foldOnlyLines is used when you can't fold a part of the line. `

Reply:
I've added code to handle it - if I understand it correctly